src/GoTrueClient.ts

```diff
@@ -216,78 +219,82 @@ export default class GoTrueClient {
+   * @param options Valid options for password sign-ins.
    */
-  async signIn(
-    { email, phone, password, refreshToken, provider, oidc }: UserCredentials,
-    options: {
-      redirectTo?: string
-      shouldCreateUser?: boolean
-      scopes?: string
-      captchaToken?: string
-      queryParams?: { [key: string]: string }
-    } = {}
-  ): Promise<
+  async signInWithPassword({
+    email,
+    phone,
+    password,
+    options = {},
+  }: SignInWithPasswordCredentials): Promise<
     | {
         session: Session | null
         user: User | null
-        provider?: Provider
         url?: string | null
         error: null
       }
     | {
         session: Session | null
         user: User | null
-        provider?: Provider
         url?: string | null
         error: AuthError
       }
   > {
     try {
       this._removeSession()
 
-      if (email && !password) {
-        const { error } = await this.api.sendMagicLinkEmail(email, {
-          redirectTo: options.redirectTo,
-          shouldCreateUser: options.shouldCreateUser,
-          captchaToken: options.captchaToken,
-        })
-        return { user: null, session: null, error }
-      }
       if (email && password) {
         return this._handleEmailSignIn(email, password, {
-          redirectTo: options.redirectTo,
-        })
-      }
-      if (phone && !password) {
-        const { error } = await this.api.sendMobileOTP(phone, {
-          shouldCreateUser: options.shouldCreateUser,
           captchaToken: options.captchaToken,
         })
-        return { user: null, session: null, error }
       }
       if (phone && password) {
         return this._handlePhoneSignIn(phone, password)
       }
-      if (refreshToken) {
-        // currentSession and currentUser will be updated to latest on _callRefreshToken using the passed refreshToken
-        const { error } = await this._callRefreshToken(refreshToken)
-        if (error) throw error
-
-        return {
-          user: this.currentUser,
-          session: this.currentSession,
-          error: null,
-        }
+      throw new Error(`You must provide either an email or phone number and a password.`)
+    } catch (error) {
+      if (isAuthError(error)) {
+        return { user: null, session: null, error }
+      }
+
+      throw error
+    }
+  }
+
+  /**
+   * Log in an existing user via a third-party provider.
+   * @type SignInWithOAuthCredentials
+   * @param provider One of the providers supported by GoTrue.
+   * @param oidc See OpenIdCredentials type specification
+   * @param redirectTo A URL to send the user to after they are confirmed (OAuth logins only).
+   * @param scopes A space-separated list of scopes granted to the OAuth application.
+   * @param queryParams An object of query params 
+   */
+  async signInWithOAuth(
+    { provider, oidc, options = {} }: SignInWithOAuthCredentials,
+  ): Promise<
+    | {
```

Comment:
Can we create proper interfaces for the responses instead of writing them out like this inside of each method?

Reply:
added an [`AuthResponse`](https://github.com/supabase/gotrue-js/pull/304/files#diff-b8b77f5be18cf56dca425b3a5b8e9d2e754dd37fe0e3612b95cd4e9bccda08a5R29-R37) type